core/controllers/ingest.py

```diff
@@ -162,72 +146,14 @@ def populate_db(workbook: dict[str, pd.DataFrame], mappings: tuple[DataMapping])
-        .join(Submission)
-        .filter(Programme.programme_id == programme_id)
-        .filter(Submission.reporting_round == reporting_round)
-        .first()
-    )
-    # if added before or during this round, get the programme entity to merge (update)
-    programme_exists_previous_round = (
-        Programme.query.join(Project)
-        .join(Submission)
-        .filter(Programme.programme_id == programme_id)
-        .filter(Submission.reporting_round <= reporting_round)
-        .first()
-    )
-
-    if programme_exists_same_round:
-        matching_project = None
-        for project in programme_exists_same_round.projects:
-            if project.submission.reporting_round == reporting_round:
-                matching_project = project
-                break
-        if matching_project:
-            # get id of submission to replace and re-use it
-            submission_id = matching_project.submission.submission_id
-            # submission instance to remove
-            submission_to_del = matching_project.submission
-            # drop submission, all children should be dropped via cascade
-            Submission.query.filter_by(id=submission_to_del.id).delete()
-            db.session.flush()
-
-    else:
-        submission_id = next_submission_id(reporting_round)
+    submission_id = get_submission_id(workbook)
 
     for mapping in mappings:
-        worksheet = workbook[mapping.table]
-        if "Submission ID" in mapping.column_mapping:
-            worksheet["Submission ID"] = submission_id
-        models = mapping.map_data_to_models(worksheet)
-
-        if mapping.table == "Programme_Ref" and programme_exists_previous_round:
-            # Set incoming model pk to match existing DB row pk (this record will then be updated).
-            programme_to_merge = models[0]
-            programme_to_merge.id = programme_exists_previous_round.id
-            db.session.merge(programme_to_merge)  # There can only be 1 programme per ingest.
-            continue
-
-        if mapping.table == "Organisation_Ref":
-            organisation_exists = Organisation.query.filter(
-                Organisation.organisation_name == models[0].organisation_name
-            ).first()
-            # If this org already in DB, merge to re-use pk, otherwise use add_all as per loop continuation
-            if organisation_exists:
-                org_to_merge = models[0]
-                org_to_merge.id = organisation_exists.id
-                db.session.merge(org_to_merge)  # There can only be 1 org per ingest.
-                continue
-
-        # for outcome and output dim (ref) data, if record already exists, do nothing.
-        if mapping.table in ["Outputs_Ref", "Outcome_Ref"]:
-            models = get_outcomes_outputs_to_insert(mapping, models)
-
-        db.session.add_all(models)
+        load_function = TABLE_TO_LOAD_FUNCTION[mapping.table]
+        if load_function.__name__ == "generic_load":
```

Comment:
why do you need to use.__name__? can't you just compare functions
```if load_function == generic_load:```

Reply:
since generic_load is basicly the default, you could consider only storing the entries for the other cases:
```
TABLE_TO_LOAD_FUNCTION = {   
    "Organisation_Ref": load_organisation_ref,
    "Programme_Ref": load_programme_ref,
    "Programme Progress": generic_load,
    "Funding": generic_load,
    "Outputs_Ref": load_outputs_outcomes_ref,
    "Outcome_Ref": load_outputs_outcomes_ref,
 }
 ```

 then do something like:
 
 ```
 if load_function := TABLE_TO_LOAD_FUNCTION.get(mapping.table):
     load_function() 
 else:
     generic_load()
```    
 But if you don't think this is an improvement then no worries

Reply:
ingest.py has no concept of ```generic_load``` so it would need to be imported. This is probably a quicker and less complicated look-up

Reply:
Ah yeah, those functions are stored in a separate module. This way is probably best then